pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnxThrottleTracker.java

```diff
@@ -18,125 +18,439 @@
+         * </ul>
+         */
+        Dropped
     }
 
     /**
-     * See {@link Producer#incrementThrottleCount()} for documentation.
+     * Checks if the connection is currently throttled by any throttle type.
+     *
+     * <p>This method examines all throttle type states and returns {@code true}
+     * if any of them are active (counter > 0).
+     *
+     * @return {@code true} if any throttling condition is active, {@code false} otherwise
      */
-    public void incrementThrottleCount() {
-        int currentThrottleCount = THROTTLE_COUNT_UPDATER.incrementAndGet(this);
-        if (currentThrottleCount == 1) {
-            changeAutoRead(false);
+    private boolean hasThrottled() {
+        for (int stat : states) {
+            if (stat > 0) {
+                return true;
+            }
         }
+        return false;
     }
 
     /**
-     * See {@link Producer#decrementThrottleCount()} for documentation.
+     * Returns the total count of active throttling conditions across all types.
+     *
+     * <p>This method sums up all the individual counters for each throttle type,
+     * providing a measure of the overall throttling pressure on the connection.
+     * For reentrant types, this includes the full counter value (not just 0 or 1).
+     *
+     * @return the total number of active throttling conditions
      */
-    public void decrementThrottleCount() {
-        int currentThrottleCount = THROTTLE_COUNT_UPDATER.decrementAndGet(this);
-        if (currentThrottleCount == 0) {
-            changeAutoRead(true);
+    public int throttledCount() {
+        int i = 0;
+        for (int stat : states) {
+            i += stat;
         }
+        return i;
     }
 
-    private void changeAutoRead(boolean autoRead) {
-        if (isChannelActive()) {
+    /**
+     * Marks the connection as throttled for the specified throttle type.
+     *
+     * <p>This method activates throttling for the given type and may pause the connection's
+     * request processing if this is the first active throttling condition. For reentrant
+     * types ({@link ThrottleType#TopicPublishRate} and {@link ThrottleType#ResourceGroupPublishRate}),
+     * this increments the counter. For non-reentrant types, this sets the state to active.
+     *
+     * <p>If the connection transitions from unthrottled to throttled, this method will
+     * set the Netty channel's auto-read to {@code false}, effectively pausing incoming
+     * request processing.
+     *
+     * <p>Metrics are automatically recorded to track throttling events and connection state changes.
+     *
+     * @param type the type of throttling condition to activate
+     * @throws IllegalArgumentException if type is null
+     *
+     * @see #unmarkThrottled(ThrottleType)
+     * @see ThrottleType
+     */
+    public void markThrottled(ThrottleType type) {
+        ThrottleRes res = doMarkThrottled(type);
```

Comment:
At the beginning of the method, could we add a check here to ensure that this is executed within the event loop of the connection?

something like
```
assert serverCnx.ctx().executor().inEventLoop() : "This method should be called in serverCnx.ctx().executor()";
```

Another possibility would be to always accept the call and if it's not in the eventloop, then run the call to the event loop.
```
if (!serverCnx.ctx().executor().inEventLoop()) {
   serverCnx.ctx().executor().execute(() -> markThrottled(type));
}
```